File: as/src/androidTest/java/com/trikset/gamepad/MainWindowTests.java
package com.trikset.gamepad;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import androidx.test.espresso.UiController;
import androidx.test.espresso.ViewAction;
import androidx.test.espresso.action.MotionEvents;
import androidx.test.filters.LargeTest;
import androidx.test.rule.ActivityTestRule;

import org.hamcrest.Matcher;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
import org.junit.runners.Parameterized.Parameters;

import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;


@RunWith(Enclosed.class)
public class MainWindowTests {
    @LargeTest
    @RunWith(Parameterized.class)
    public static class SquareButtonTest {
        @Rule
        public final ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class);

        @Parameters
        public static Collection<Object[]> data() {
            return Arrays.asList(new Object[][]{{R.id.leftPad, "1"}, {R.id.rightPad, "2"}});
        }

        @Parameter
        public int currentPadId;
        @Parameter(1)
        public String currentPadName;

        @Before
        public void initNetworkSettings() {
            SharedPreferences preferences =
                    PreferenceManager.getDefaultSharedPreferences(mActivityTestRule.getActivity());
            SharedPreferences.Editor preferenceEditor = preferences.edit();

            preferenceEditor.putString(SettingsFragment.SK_HOST_ADDRESS, DummyServer.IP);
            preferenceEditor.putString(
                    SettingsFragment.SK_HOST_PORT,
                    Integer.toString(DummyServer.DEFAULT_PORT));
            // In order not to receive keep-alive messages
            preferenceEditor.putString(SettingsFragment.SK_KEEPALIVE, "100000000");

            preferenceEditor.commit();
        }

        @Test
        public void squareButtonsShouldHandleCircularTapsCorrectly() {
            DummyServer server = new DummyServer();

            onView(withId(currentPadId)).perform(movingTap());
            server.stopListening();

            Iterator<String> messages = server.getReceivedMessages().iterator();
            while (messages.hasNext()) {
                final String current = messages.next();

                if (messages.hasNext()) {
                    assertNotEquals(String.format("pad %s up", currentPadName), current);

                    String[] splitCommand = current.split(" ");
                    assertEquals(4, splitCommand.length);
                    assertEquals("pad", splitCommand[0]);
                    assertEquals(currentPadName, splitCommand[1]);

                    int x = Integer.parseInt(splitCommand[2]);
                    int y = Integer.parseInt(splitCommand[3]);
                    int radius = (int) Math.sqrt(x * x + y * y);
                    assertTrue((radius > 40) && (radius < 60));
                } else {
                    assertEquals(String.format("pad %s up", currentPadName), current);
                }
            }
        }

        @Test
        public void squareButtonsShouldHandleDiagonalTapsCorrectly() {
            DummyServer server = new DummyServer();

            onView(withId(currentPadId)).perform(diagonalTap());
            server.stopListening();

            Iterator<String> messages = server.getReceivedMessages().iterator();
            int currentIndex = 0;
            while (messages.hasNext()) {
                String current = messages.next();

                if (!messages.hasNext()) {
                    assertEquals(String.format("pad %s up", currentPadName), current);
                    break;
                }

                assertNotEquals(String.format("pad %s up", currentPadName), current);

                String[] splitCommand = current.split(" ");
                assertEquals(4, splitCommand.length);
                assertEquals("pad", splitCommand[0]);
                assertEquals(currentPadName, splitCommand[1]);

                int x = Integer.parseInt(splitCommand[2]);
                int y = Integer.parseInt(splitCommand[3]);

                assertTrue(Math.abs(-100 + currentIndex * 200 / 10 - x) <= 25);
                assertTrue(Math.abs(100 - currentIndex * 200 / 10 - y) <= 25);

                ++currentIndex;
            }
        }

        private ViewAction diagonalTap() {
            return new ViewAction() {
                final static int tapSegmentCount = 10;
                final float[] tapPrecision = new float[]{1f, 1f};

                @Override
                public Matcher<View> getConstraints() {
                    return isDisplayed();
                }

                @Override
                public String getDescription() {
                    return "Diagonal tap from top left corner to bottom right corner";
                }

                @Override
                public void perform(UiController uiController, View view) {
                    int[] topLeftCoords = new int[2];
                    view.getLocationOnScreen(topLeftCoords);

                    final float[] startCoords = {topLeftCoords[0] + tapPrecision[0], topLeftCoords[1] - tapPrecision[1]};
                    MotionEvent tap = MotionEvents.sendDown(uiController, startCoords, tapPrecision).down;
                    uiController.loopMainThreadUntilIdle();
                    try {
                        for (int i = 1; i < tapSegmentCount; ++i) {
                            final float[] currentCoords = {startCoords[0] + (float) i * (view.getWidth()) / tapSegmentCount, startCoords[1] - (float) i * (view.getHeight()) / tapSegmentCount,};
                            if (!MotionEvents.sendMovement(uiController, tap, currentCoords)) {
                                MotionEvents.sendCancel(uiController, tap);
                                break;
                            }
                        }
                        if (!MotionEvents.sendUp(uiController, tap)) {
                            MotionEvents.sendCancel(uiController, tap);
                        }
                    } finally {
                        tap.recycle();
                    }
                }
            };
        }

        private ViewAction movingTap() {
            return new ViewAction() {
                final static int tapSegmentCount = 10;
                final float[] tapPrecision = new float[]{1f, 1f};

                @Override
                public Matcher<View> getConstraints() {
                    return isDisplayed();
                }

                @Override
                public String getDescription() {
                    return "Circular press around the starting point";
                }

                @Override
                public void perform(UiController uiController, View view) {
                    final int tapRadius = view.getWidth() / 4;

                    int[] topLeftCoords = new int[2];
                    view.getLocationOnScreen(topLeftCoords);
                    int[] centerCoords = {topLeftCoords[0] + view.getWidth() / 2, topLeftCoords[1] + view.getWidth() / 2};

                    float[] startCoords = new float[]{centerCoords[0] + tapRadius, centerCoords[1]};
                    MotionEvent tap = MotionEvents.sendDown(uiController, startCoords, tapPrecision).down;
                    uiController.loopMainThreadForAtLeast(100);
                    for (int i = 1; i <= tapSegmentCount; ++i) {
                        double currentAngle = 2 * i * Math.PI / tapSegmentCount;
                        float[] currentCoords = {centerCoords[0] + tapRadius * (float) Math.cos(currentAngle), centerCoords[1] + tapRadius * (float) Math.sin(currentAngle)};

                        MotionEvents.sendMovement(uiController, tap, currentCoords);
                        uiController.loopMainThreadForAtLeast(50);
                    }
                    MotionEvents.sendUp(uiController, tap);
                }
            };
        }
    }

    @LargeTest
    @RunWith(JUnit4.class)
    public static class MagicButtonsTests {
        @Rule
        public final ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class);

        @Before
        public void initNetworkSettings() {
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(mActivityTestRule.getActivity());
            SharedPreferences.Editor preferenceEditor = preferences.edit();

            preferenceEditor.putString(SettingsFragment.SK_HOST_ADDRESS, DummyServer.IP);
            preferenceEditor.putString(SettingsFragment.SK_HOST_PORT, Integer.toString(DummyServer.DEFAULT_PORT));
            // In order not to receive keep-alive messages
            preferenceEditor.putString(SettingsFragment.SK_KEEPALIVE, "100000000");

            preferenceEditor.commit();
        }

        @Test
        public void magicButtonsShouldSendCorrectCommands() {
            DummyServer server = new DummyServer();

            onView(withId(R.id.buttons)).perform(pressButtons());
            server.stopListening();

            Iterator<String> messages = server.getReceivedMessages().iterator();
            for (int i = 1; i <= 5; ++i) {
                assertTrue(messages.hasNext());
                String currentMessage = messages.next();

                assertEquals(String.format("btn %d down", i), currentMessage);
            }
            assertFalse(messages.hasNext());
        }

        private ViewAction pressButtons() {
            return new ViewAction() {
                final float[] tapPrecision = {1f, 1f};

                @Override
                public Matcher<View> getConstraints() {
                    return isDisplayed();
                }

                @Override
                public String getDescription() {
                    return "Clicks on each of magic buttons";
                }

                @Override
                public void perform(UiController uiController, View view) {
                    final ViewGroup buttons = (ViewGroup) view;

                    for (int i = 0; i <= 4; ++i) {
                        final View currentButton = buttons.getChildAt(i);

                        int[] buttonTopLeftCoords = new int[2];
                        currentButton.getLocationOnScreen(buttonTopLeftCoords);
                        float[] clickCoords = {buttonTopLeftCoords[0] + (float) currentButton.getWidth() / 2, buttonTopLeftCoords[1] + (float) currentButton.getHeight() / 2};

                        MotionEvent tap = MotionEvents.sendDown(uiController, clickCoords, tapPrecision).down;
                        uiController.loopMainThreadForAtLeast(50);
                        MotionEvents.sendUp(uiController, tap);
                    }
                }
            };
        }
    }
}

File: app/src/androidTest/kotlin/com/trikset/gamepad2/MainWindowTests.kt
package com.trikset.gamepad2

import android.view.View
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.action.MotionEvents
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withContentDescription
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.filters.LargeTest
import java.util.Locale
import org.hamcrest.Matcher
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameter
import org.junit.runners.Parameterized.Parameters

@RunWith(Enclosed::class)
class MainWindowTests {

  private companion object {
    val tapPrecision = floatArrayOf(1f, 1f)
  }

  @LargeTest
  @RunWith(Parameterized::class)
  class SquareButtonTest {
    @get:Rule val mActivityTestRule = FocusAwareActivityTestRule(MainActivity::class.java)

    @Parameter @JvmField var currentPadId: Int = 0

    @Parameter(1) @JvmField var currentPadName: String = ""

    @Before
    fun initNetworkSettings() {
      initNetworkSettings(mActivityTestRule.activity)
    }

    @Test
    fun squareButtonsShouldHandleCircularTapsCorrectly() {
      val server = DummyServer()
      onView(withId(currentPadId)).perform(movingTap())
      // The trailing 'up' is sent asynchronously (pad -> SenderService executor ->
      // socket); stopListening() drops any line that lands after its flag is set,
      // so bounded-await the up before stopping (hit 2026-08-11 on a physical phone).
      assertTrue(
          "expected 'pad $currentPadName up', received so far: ${server.receivedMessages}",
          server.awaitMessage(String.format(Locale.ROOT, "pad %s up", currentPadName), 30_000),
      )
      server.stopListening()

      assertPadCommands(server.receivedMessages, currentPadName) { x, y ->
        val radius = Math.sqrt((x * x + y * y).toDouble())
        assertTrue(radius > 40 && radius < 60)
      }
    }

    @Test
    fun squareButtonsShouldHandleDiagonalTapsCorrectly() {
      val server = DummyServer()
      onView(withId(currentPadId)).perform(diagonalTap())
      // Same async-'up' race as the circular tap: bounded-await before stopping.
      assertTrue(
          "expected 'pad $currentPadName up', received so far: ${server.receivedMessages}",
          server.awaitMessage(String.format(Locale.ROOT, "pad %s up", currentPadName), 30_000),
      )
      server.stopListening()

      var currentIndex = 0
      assertPadCommands(server.receivedMessages, currentPadName) { x, y ->
        // The DOWN starts at command (-80,80) (Samsung top strip is a dead zone for
        // the (100,100) corner), so the expected diagonal baseline is (-80,80), not
        // (-100,100); ±25 absorbs the int-truncation rounding in the received values.
        assertTrue(Math.abs(-80 + currentIndex * 200 / 10 - x) <= 25)
        assertTrue(Math.abs(80 - currentIndex * 200 / 10 - y) <= 25)
        ++currentIndex
      }
    }

    /**
     * Asserts [messages] is a sequence of `pad <name> x y` moves (validated by [validateMove])
     * ending with `pad <name> up`.
     */
    private fun assertPadCommands(
        messages: List<String>,
        padName: String,
        validateMove: (x: Int, y: Int) -> Unit,
    ) {
      val iterator = messages.iterator()
      while (iterator.hasNext()) {
        val current = iterator.next()
        val up = String.format(Locale.ROOT, "pad %s up", padName)
        if (iterator.hasNext()) {
          assertNotEquals(up, current)
          val splitCommand = current.split(" ")
          assertEquals(4, splitCommand.size)
          assertEquals("pad", splitCommand[0])
          assertEquals(padName, splitCommand[1])
          validateMove(splitCommand[2].toInt(), splitCommand[3].toInt())
        } else {
          assertEquals(up, current)
        }
      }
    }

    private fun diagonalTap(): ViewAction {
      return object : ViewAction {
        private val tapSegmentCount = 10

        override fun getConstraints(): Matcher<View> = isDisplayed()

        override fun getDescription(): String =
            "Diagonal tap from top left corner to bottom right corner"

        override fun perform(uiController: UiController, view: View) {
          val topLeftCoords = IntArray(2)
          view.getLocationOnScreen(topLeftCoords)
          // Invert TouchPadController's mapping (command = 200*1.15*(frac-0.5)) so every
          // touch point lands INSIDE the pad. The old path started 1px above the pad's
          // top edge and moved further up-right, so the DOWN never reached the pad and
          // the test only passed vacuously on an empty message list (hit 2026-08-11,
          // physical phone; the bounded await exposed the missing 'up'). Start at command
          // (-80,80) instead of (-100,100): on Samsung phones the top ~125px of the
          // screen is a dead strip (status/gesture area) and a DOWN at screen y≈84 is
          // swallowed; the assertion's ±25 tolerance absorbs the constant 20 offset.
          fun touchPoint(commandX: Int, commandY: Int): FloatArray {
            val xFrac = 0.5 + commandX / COMMAND_SCALE
            val yFrac = 0.5 - commandY / COMMAND_SCALE
            return floatArrayOf(
                topLeftCoords[0] + (xFrac * view.width).toFloat(),
                topLeftCoords[1] + (yFrac * view.height).toFloat(),
            )
          }
          val tap = MotionEvents.sendDown(uiController, touchPoint(-80, 80), tapPrecision).down
          uiController.loopMainThreadUntilIdle()
          try {
            for (i in 1 until tapSegmentCount) {
              val currentCoords =
                  touchPoint(
                      -100 + i * 200 / tapSegmentCount,
                      100 - i * 200 / tapSegmentCount,
                  )
              if (!MotionEvents.sendMovement(uiController, tap, currentCoords)) {
                MotionEvents.sendCancel(uiController, tap)
                break
              }
            }
            if (!MotionEvents.sendUp(uiController, tap)) {
              MotionEvents.sendCancel(uiController, tap)
            }
          } finally {
            tap.recycle()
          }
        }
      }
    }

    private fun movingTap(): ViewAction {
      return object : ViewAction {
        private val tapSegmentCount = 10

        override fun getConstraints(): Matcher<View> = isDisplayed()

        override fun getDescription(): String = "Circular press around the starting point"

        override fun perform(uiController: UiController, view: View) {
          val tapRadius = view.width / 4
          val topLeftCoords = IntArray(2)
          view.getLocationOnScreen(topLeftCoords)
          val centerCoords =
              intArrayOf(
                  topLeftCoords[0] + view.width / 2,
                  topLeftCoords[1] + view.width / 2,
              )
          val startCoords =
              floatArrayOf(
                  centerCoords[0] + tapRadius.toFloat(),
                  centerCoords[1].toFloat(),
              )
          val tap = MotionEvents.sendDown(uiController, startCoords, tapPrecision).down
          uiController.loopMainThreadForAtLeast(100)
          for (i in 1..tapSegmentCount) {
            val currentAngle = 2 * i * Math.PI / tapSegmentCount
            val currentCoords =
                floatArrayOf(
                    centerCoords[0] + tapRadius * Math.cos(currentAngle).toFloat(),
                    centerCoords[1] + tapRadius * Math.sin(currentAngle).toFloat(),
                )
            MotionEvents.sendMovement(uiController, tap, currentCoords)
            uiController.loopMainThreadForAtLeast(50)
          }
          MotionEvents.sendUp(uiController, tap)
        }
      }
    }

    companion object {
      // TouchPadController maps touch fraction -> command as 200*1.15*(frac-0.5);
      // invert that here so the injected points hit the pad (constants are private
      // there, so mirror them for the test).
      const val COMMAND_SCALE = 200.0 * 1.15

      @JvmStatic
      @Parameters
      fun data(): Collection<Array<Any>> =
          listOf(arrayOf(R.id.leftPad, "1"), arrayOf(R.id.rightPad, "2"))
    }
  }

  @LargeTest
  @RunWith(JUnit4::class)
  class MagicButtonsTests {
    // The magic-button row renders at the configured count during onCreate; set it
    // to 5 before the activity launches so the full row (buttons 1..5) is asserted.
    @get:Rule
    val mActivityTestRule =
        object : FocusAwareActivityTestRule<MainActivity>(MainActivity::class.java) {
          override fun beforeActivityLaunched() {
            androidx.preference.PreferenceManager.getDefaultSharedPreferences(
                    androidx.test.platform.app.InstrumentationRegistry.getInstrumentation()
                        .targetContext
                )
                .edit()
                .putInt(SettingsFragment.SK_MAGIC_BUTTON_COUNT, 5)
                .commit()
          }
        }

    private val defaultGlyphs = listOf("▲", "■", "●", "✕", "◆")

    @Before
    fun initNetworkSettings() {
      initNetworkSettings(mActivityTestRule.activity)
    }

    @Test
    @Throws(InterruptedException::class)
    fun magicButtonsShouldSendCorrectCommands() {
      val server = DummyServer()
      for (i in 1..5) {
        // The description is locale-dependent word + "· <glyph>" (e.g. "Кнопка 1 · ▲").
        // Match by the glyph suffix, which is the same in every locale.
        onView(withContentDescription(org.hamcrest.Matchers.endsWith("· ${defaultGlyphs[i - 1]}")))
            .perform(performClickAction())
        // The listener fires an async send (connect + write on a real executor); bounded-await
        // each command so the socket write lands (TESTING.md: never a bare assert on
        // server-received content).
        assertTrue(
            "expected 'btn $i down', received so far: ${server.receivedMessages}",
            server.awaitMessage(String.format(Locale.ROOT, "btn %d down", i), 30_000),
        )
      }
      server.stopListening()

      val messages = server.receivedMessages.iterator()
      for (i in 1..5) {
        assertTrue(messages.hasNext())
        val currentMessage = messages.next()
        assertEquals(String.format(Locale.ROOT, "btn %d down", i), currentMessage)
      }
      assertFalse(messages.hasNext())
    }

    /** A [ViewAction] that calls [View.performClick] directly (bypasses touch injection). */
    private fun performClickAction(): ViewAction {
      return object : ViewAction {
        override fun getConstraints(): Matcher<View> = isDisplayed()

        override fun getDescription(): String = "performClick"

        override fun perform(uiController: UiController, view: View) {
          view.performClick()
        }
      }
    }
  }
}
